core/src/word_lock.rs

```diff
@@ -189,7 +189,7 @@ impl WordLock {
             match self.state.compare_exchange_weak(
                 state,
                 state | QUEUE_LOCKED_BIT,
-                Ordering::Acquire,
+                Ordering::AcqRel,
```

Comment:
This (and the changes below) doesn't make sense: there's nothing to release. The lock itself was already released in `unlock` with release ordering.

However I am noticing that the `state` load at the start of this function should be `Acquire` since it needs to happen before reading the state entries. Perhaps fixing this makes tsan happy?

Reply:
> there's nothing to release. The lock itself was already released in unlock with release ordering.

Ah, my bad, yeah I see, I didn't look that closely, and thought this was still trying to unlock not to wake up threads 😅

> However I am noticing that the state load at the start of this function should be Acquire since it needs to happen before reading the state entries.

So, this doesn't make sense to me, since it enters a loop that can only be exited by either returning (in which case the data in behind the ptr inside state is not read) or by passing the cmpxchg, which gives the Acquire.

... And in fact, yeah, I was mistaken that this needed AcqRel — this one was actually fine as Acquire. (Sorry...)

Actually, I think `unlock_slow` might be entirely fine, and `lock_slow` is the only problem. It looks like the other two changes in that are probably caused by TSan not understanding how fences work in all cases (I had heard this was fixed, but it seems it is didn't get merged into LLVM yet, and has stalled. Although, in June there was some progress, , so maybe someday soon).

Regardless, if I apply the normal workaround for acquire fences under TSan (where you replace a fence with a discarded acquire load), then L252/L233 can both use Release.

(Just to make extra-sure, I verified that the change on L157 is needed with or without the tsan fence fix).

I'll have an updated patch shortly.